torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -604,7 +613,20 @@ def _pre_backward_hook(*unused: Any) -> None:
 
             # All-gather full parameters, moving them to compute device if
             # necessary.
+            # Always wait for all_gather before rebuilding full params, just
+            # in case full params are prefetched, if full params are not prefetched,
+            # it is no-op to wait for all gather stream
+            torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
             self._rebuild_full_params()
+            # Wait for all_gather to finish before computation
+            torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
+
+            # Prefetch previous layer's full params in backward pass
+            if (
+                self._fsdp_graph_order is not None
+                and self._my_fsdp_idx_in_graph is not None and self._my_fsdp_idx_in_graph > 0
+            ):
+                self._fsdp_graph_order[self._my_fsdp_idx_in_graph - 1]._rebuild_full_params()  # type: ignore[operator]
```

Comment:
what happens to the existing `self._rebuild_full_params()` calls? Is there a risk we call them redundantly for the same module, or is this handled somehow?

Reply:
it is handled inside _rebuild_full_params()